Checkin app week 3

webbower opened the pull request:
* Finish designing Redux state and writing creation actions
* Refactor reducer to use immer for deep reducer state updates
* Install tools for React development
* Integrate Redux store into app
* Build read-only Users and Teams pages with tests and dummy data
* Convert app to Next.js and remove legacy config

Commit:
Refactor checkin tests based on 1:1 feedback

* Remove zip logic from `createTestState()` for checkins and tasks
* Convert literal test data values into variables

Commit:
Team action and test data creators use ID in default name

Commit:
Create Redux state for Blockers with tests

Commit:
Install Immer and refactor reducer to use Immer

Commit:
Install and set up React for UI development

* Installed Babel, Webpack, React, PropTypes, Redux, and HMR
* Updated ESLint config for React
* Updated test script to support testing React components
* Added starter files for serving React UI

Commit:
Upgrade ESLint and ESLint import plugin

Upgraded to resolve dependency warning

Commit:
Integrate Redux store and Provider into App

Commit:
Build read-only Users Page with tests

* Create Heading component with tests

Commit:
Add React Router and integrate Users page into app

Commit:
Build read-only Teams Page with tests and integrate into app

ericelliott commented on src/App.jsx:
```diff
@@ -0,0 +1,39 @@
+              </li>
+            </ul>
+          </nav>
+          <Switch>
```
Switch to Next.js and you won't need to write routing logic for 99% of the app. It's handled by convention by dropping files into the `/pages` folder.

webbower replied:
Will address this as part of migrating to Next.js in another branch/PR

ericelliott commented on src/App.jsx:
```diff
@@ -0,0 +1,39 @@
+class App extends React.Component {
+  render() {
+    return (
+      <Router>
```
You probably should not have an app component that actually renders nav in your app file.

webbower replied:
Will address this as part of migrating to Next.js in another branch/PR

ericelliott commented on src/App.jsx:
```diff
@@ -0,0 +1,39 @@
+  }
+}
+
+export default hot(module)(App);
```
Config stuff like this can be composed into your app using a page HOC that wraps around every page.

webbower replied:
The `hot` HOC will be gone, but will create a `page` HOC with Redux `connect` and such after migration to Next.js

ericelliott commented on src/components/teams-page/teams-page.test.jsx:
```diff
@@ -0,0 +1,64 @@
+      given: 'an array of teams',
+      should: 'display the list of teams',
+      actual: [
+        $('.teams-list').text().trim().includes('12345'),
```
You don't have to trim if you're using includes.

ericelliott commented on src/redux/checkin.js:
```diff
@@ -98,6 +108,31 @@ export const addTask = ({
+});
+
+// Selectors
+export const getUsersList = (state) => Object.entries(state.users).map(([, user]) => user);
```
This is the same as using Object.values, but more code.

ericelliott commented on src/redux/checkin.js:
```diff
@@ -98,6 +108,31 @@ export const addTask = ({
+  Object.entries(state.teams).map(([, team]) => ({
+    id: team.id,
+    name: team.name,
+    users: team.users.map((id) => (state.users[id] || {}).name),
```
Maybe turn this into a reusable denormalize abstraction.

webbower replied:
Created `joinOn` utility

ericelliott commented on src/redux/checkin.test.js:
```diff
@@ -301,35 +325,246 @@ describe('checkin: adding tasks', async (assert) => {
+  assert({
+    given: 'adding multiple tasks to existing checkins by existing users',
+    should: 'add the tasks to the state',
+    actual: [
```
For complex setup, try the Arrange, Act, Assert structure. Arrange first, then act, then assert last. That will make your assertion and the intention of the test much more clear.

webbower replied:
I ended up combining the Arrange and Act portions since it seemed like separating them wouldn't be of much benefit.

ericelliott commented on src/redux/checkin.test.js:
```diff
@@ -301,35 +325,246 @@ describe('checkin: adding tasks', async (assert) => {
+
+  assert({
+    given: 'teams in state',
+    should: 'return an array or teams',
```
array of teams?

webbower replied:
Also fixed an "array or users" from the `getUsersList` test that this was duplicated from.

ericelliott reviewed (commented):
This is looking great!  Check out autodux to remove some of the redux boilerplate. Bonus points: Autodux + Immer = magic.

Commit:
Address PR feedback

- Fixed typos
- Removed `.trim()` in some tests
- Convert Redux selectors to point-free style
- Moved large data structures out of `assert` calls
- Add `pipe`, `prop`, and `joinOn` utilities with tests

Commit:
Migrate working parts of app to Next.js

Previous implementation code will be removed in next commit

Commit:
Remove previous app config

Legacy app config removed. Pure Next.js app now
